Hibernate ORM user types should be registered for reflection

From Zulip:
The attribute in my class is:
```
@Type(type = "jsonb")
private String payload;
```

I have declared the jsonb type:

```
@TypeDefs({
        @TypeDef(name = "jsonb", typeClass = RawJsonType.class)
})
```

```
Caused by: org.hibernate.MappingException: Could not determine type for: br.com.cresolconfederacao.idempotencia.hibernate.RawJsonType, at table: Registro, for columns: [org.hibernate.mapping.Column(payload)]
```

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java
@@ -1,0 +1,84 @@
+package io.quarkus.hibernate.orm.deployment;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.hibernate.annotations.Type;
+import org.hibernate.annotations.TypeDef;
+import org.hibernate.annotations.TypeDefs;
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationValue;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.IndexView;
+
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
+
+public final class HibernateUserTypeProcessor {
+    private static final String TYPE_VALUE = "type";
+    private static final String TYPE_CLASS_VALUE = "typeClass";
+    private static final DotName TYPE = DotName.createSimple(Type.class.getName());
+    private static final DotName TYPE_DEFINITION = DotName.createSimple(TypeDef.class.getName());
+    private static final DotName TYPE_DEFINITIONS = DotName.createSimple(TypeDefs.class.getName());
+
+    @BuildStep
+    public void build(BuildProducer<ReflectiveClassBuildItem> reflectiveClass, CombinedIndexBuildItem combinedIndexBuildItem) {
+        IndexView index = combinedIndexBuildItem.getIndex();
+
+        final Set<String> userTypes = new HashSet<>();
+
+        Collection<AnnotationInstance> typeAnnotationInstances = index.getAnnotations(TYPE);
+        Collection<AnnotationInstance> typeDefinitionAnnotationInstances = index.getAnnotations(TYPE_DEFINITION);
+        Collection<AnnotationInstance> typeDefinitionsAnnotationInstances = index.getAnnotations(TYPE_DEFINITIONS);
+
+        userTypes.addAll(getUserTypes(typeDefinitionAnnotationInstances));
+
+        for (AnnotationInstance typeDefinitionAnnotationInstance : typeDefinitionsAnnotationInstances) {
+            final AnnotationValue typeDefinitionsAnnotationValue = typeDefinitionAnnotationInstance.value();
+
+            if (typeDefinitionsAnnotationValue == null) {
+                continue;
+            }
+
+            userTypes.addAll(getUserTypes(Arrays.asList(typeDefinitionsAnnotationValue.asNestedArray())));
+        }
+
+        for (AnnotationInstance typeAnnotationInstance : typeAnnotationInstances) {
+            final AnnotationValue typeValue = typeAnnotationInstance.value(TYPE_VALUE);
+            if (typeValue == null) {
+                continue;
+            }
+
+            final String type = typeValue.asString();
+            final DotName className = DotName.createSimple(type);
+            if (index.getClassByName(className) == null) {
+                continue; // Either already registered through TypeDef annotation scanning or not present in the index
+            }
+            userTypes.add(type);
+        }
+
+        if (!userTypes.isEmpty()) {
+            reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, userTypes.toArray(new String[] {})));
+        }
+    }
+
+    private Set<String> getUserTypes(Collection<AnnotationInstance> typeDefinitionAnnotationInstances) {
+        final Set<String> userTypes = new HashSet<>();
+
+        for (AnnotationInstance typeDefAnnotationInstance : typeDefinitionAnnotationInstances) {
+            final AnnotationValue typeClassValue = typeDefAnnotationInstance.value(TYPE_CLASS_VALUE);
+            if (typeClassValue == null) {
+                continue;
+            }
+
+            final String typeClass = typeClassValue.asClass().name().toString();
+            userTypes.add(typeClass);
+        }
+
+        return userTypes;
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/jpa/AbstractCustomUserType.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/jpa/AbstractCustomUserType.java
@@ -1,0 +1,76 @@
+package io.quarkus.it.jpa;
+
+import java.io.Serializable;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Objects;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.usertype.UserType;
+
+public abstract class AbstractCustomUserType<T> implements UserType {
+    Class<T> clazz;
+
+    public AbstractCustomUserType(Class<T> clazz) {
+        this.clazz = clazz;
+    }
+
+    @Override
+    public Object nullSafeGet(ResultSet result, String[] names, SharedSessionContractImplementor session, Object owner)
+            throws SQLException {
+        return get(result, names, session, owner);
+    }
+
+    @Override
+    public void nullSafeSet(PreparedStatement preparedStatement, Object value, int index,
+            SharedSessionContractImplementor session) throws SQLException {
+        set(preparedStatement, clazz.cast(value), index, session);
+    }
+
+    protected abstract T get(ResultSet rs, String[] names, SharedSessionContractImplementor session, Object owner)
+            throws SQLException;
+
+    protected abstract void set(PreparedStatement st, T value, int index, SharedSessionContractImplementor session)
+            throws SQLException;
+
+    @Override
+    public Class returnedClass() {
+        return clazz;
+    }
+
+    @Override
+    public boolean equals(Object o, Object o1) {
+        return Objects.equals(o, o1);
+    }
+
+    @Override
+    public int hashCode(Object o) {
+        return o.hashCode();
+    }
+
+    @Override
+    public Object deepCopy(Object o) {
+        return o;
+    }
+
+    @Override
+    public boolean isMutable() {
+        return true;
+    }
+
+    @Override
+    public Serializable disassemble(Object o) {
+        return (Serializable) o;
+    }
+
+    @Override
+    public Object assemble(Serializable cached, Object o) {
+        return cached;
+    }
+
+    @Override
+    public Object replace(Object o, Object t, Object owner) {
+        return o;
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/jpa/AnimalType.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/jpa/AnimalType.java
@@ -1,0 +1,40 @@
+package io.quarkus.it.jpa;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Types;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+
+public class AnimalType extends AbstractCustomUserType<Animal> {
+
+    public AnimalType() {
+        super(Animal.class);
+    }
+
+    @Override
+    public int[] sqlTypes() {
+        return new int[] { Types.DOUBLE };
+    }
+
+    @Override
+    public Animal get(ResultSet result, String[] names, SharedSessionContractImplementor session, Object owner)
+            throws SQLException {
+        double weight = result.getDouble(names[0]);
+        Animal animal = new Animal();
+        animal.setWeight(weight);
+        return animal;
+    }
+
+    @Override
+    public void set(PreparedStatement preparedStatement, Animal value, int index, SharedSessionContractImplementor session)
+            throws SQLException {
+        if (value == null) {
+            preparedStatement.setNull(index, Types.BIGINT);
+        } else {
+            preparedStatement.setDouble(index, value.getWeight());
+        }
+    }
+
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/jpa/BigIntType.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/jpa/BigIntType.java
@@ -1,0 +1,39 @@
+package io.quarkus.it.jpa;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Types;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+
+public class BigIntType extends AbstractCustomUserType<BigInteger> {
+
+    public BigIntType() {
+        super(BigInteger.class);
+    }
+
+    @Override
+    public int[] sqlTypes() {
+        return new int[] { Types.BIGINT };
+    }
+
+    @Override
+    public BigInteger get(ResultSet rs, String[] names, SharedSessionContractImplementor session, Object owner)
+            throws SQLException {
+        BigDecimal bigDecimal = rs.getBigDecimal(names[0]);
+        return bigDecimal != null ? bigDecimal.toBigIntegerExact() : null;
+    }
+
+    @Override
+    public void set(PreparedStatement st, BigInteger value, int index, SharedSessionContractImplementor session)
+            throws SQLException {
+        if (value == null) {
+            st.setNull(index, Types.BIGINT);
+        } else {
+            st.setBigDecimal(index, new BigDecimal(String.valueOf(value)));
+        }
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomEnum.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomEnum.java
@@ -1,0 +1,6 @@
+package io.quarkus.it.jpa;
+
+public enum CustomEnum {
+    ONE,
+    TWO
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomTypeEntity.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomTypeEntity.java
@@ -1,0 +1,69 @@
+package io.quarkus.it.jpa;
+
+import java.math.BigInteger;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+import org.hibernate.annotations.Type;
+import org.hibernate.annotations.TypeDef;
+import org.hibernate.annotations.TypeDefs;
+
+@Entity
+@TypeDef(typeClass = LowerCaseCustomEnumType.class, name = "lowerCaseEnumType")
+@TypeDefs({
+        @TypeDef(typeClass = AnimalType.class, name = "animalType")
+})
+public class CustomTypeEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "eventSeq")
+    private Long id;
+
+    @Column(name = "big_integer")
+    @Type(type = "io.quarkus.it.jpa.BigIntType")
+    private BigInteger bigInteger;
+
+    @Column(name = "custom_enum")
+    @Type(type = "lowerCaseEnumType")
+    private CustomEnum customEnum;
+
+    @Column(name = "animal")
+    @Type(type = "animalType")
+    private Animal animal;
+
+    public Long getId() {
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
+    public BigInteger getBigInteger() {
+        return bigInteger;
+    }
+
+    public void setBigInteger(BigInteger bigInteger) {
+        this.bigInteger = bigInteger;
+    }
+
+    public CustomEnum getCustomEnum() {
+        return customEnum;
+    }
+
+    public void setCustomEnum(CustomEnum customEnum) {
+        this.customEnum = customEnum;
+    }
+
+    public Animal getAnimal() {
+        return animal;
+    }
+
+    public void setAnimal(Animal animal) {
+        this.animal = animal;
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/jpa/JPACustomUserTypeEndpoint.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/jpa/JPACustomUserTypeEndpoint.java
@@ -1,0 +1,33 @@
+package io.quarkus.it.jpa;
+
+import java.math.BigInteger;
+
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.transaction.Transactional;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/jpa/custom-user-types")
+public class JPACustomUserTypeEndpoint {
+
+    @Inject
+    EntityManager em;
+
+    @GET
+    @Transactional
+    @Produces(MediaType.TEXT_PLAIN)
+    public Long invokeCreation() {
+        CustomTypeEntity customTypeEntity = new CustomTypeEntity();
+        customTypeEntity.setBigInteger(BigInteger.ONE);
+        customTypeEntity.setCustomEnum(CustomEnum.ONE);
+        Animal animal = new Animal();
+        animal.setWeight(29.12);
+        customTypeEntity.setAnimal(animal);
+        em.persist(customTypeEntity);
+        em.flush();
+        return customTypeEntity.getId();
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/jpa/LowerCaseCustomEnumType.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/jpa/LowerCaseCustomEnumType.java
@@ -1,0 +1,38 @@
+package io.quarkus.it.jpa;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Types;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+
+public class LowerCaseCustomEnumType extends AbstractCustomUserType<CustomEnum> {
+
+    public LowerCaseCustomEnumType() {
+        super(CustomEnum.class);
+    }
+
+    @Override
+    public int[] sqlTypes() {
+        return new int[] { Types.VARCHAR };
+    }
+
+    @Override
+    public CustomEnum get(ResultSet result, String[] names, SharedSessionContractImplementor session, Object owner)
+            throws SQLException {
+        String val = result.getString(names[0]);
+        return val != null ? CustomEnum.valueOf(val.toUpperCase()) : null;
+    }
+
+    @Override
+    public void set(PreparedStatement preparedStatement, CustomEnum value, int index, SharedSessionContractImplementor session)
+            throws SQLException {
+        if (value == null) {
+            preparedStatement.setNull(index, Types.BIGINT);
+        } else {
+            preparedStatement.setString(index, value.name().toLowerCase());
+        }
+    }
+
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/JPAUserTypeITCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/JPAUserTypeITCase.java
@@ -1,0 +1,8 @@
+package io.quarkus.it.main;
+
+import io.quarkus.test.junit.SubstrateTest;
+
+@SubstrateTest
+public class JPAUserTypeITCase extends JPAUserTypeTestCase {
+
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/JPAUserTypeTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/JPAUserTypeTestCase.java
@@ -1,0 +1,17 @@
+package io.quarkus.it.main;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.RestAssured;
+
+@QuarkusTest
+public class JPAUserTypeTestCase {
+
+    @Test
+    public void testCustomUserTypes() {
+        RestAssured.when().get("/jpa/custom-user-types").then()
+                .statusCode(200);
+    }
+
+}

